File: app/src/main/java/com/lifedawn/bestweather/notification/daily/viewmodel/DailyNotificationViewModel.java
package com.lifedawn.bestweather.notification.daily.viewmodel;

import android.app.Application;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

import com.lifedawn.bestweather.commons.enums.LocationType;
import com.lifedawn.bestweather.commons.enums.WeatherProviderType;
import com.lifedawn.bestweather.notification.daily.DailyNotificationHelper;
import com.lifedawn.bestweather.notification.daily.DailyPushNotificationType;
import com.lifedawn.bestweather.room.callback.DbQueryCallback;
import com.lifedawn.bestweather.room.dto.DailyPushNotificationDto;
import com.lifedawn.bestweather.room.dto.FavoriteAddressDto;
import com.lifedawn.bestweather.room.repository.DailyPushNotificationRepository;

import java.time.LocalTime;
import java.util.List;

public class DailyNotificationViewModel extends AndroidViewModel {
	private final DailyPushNotificationRepository repository = DailyPushNotificationRepository.getINSTANCE();
	public final LiveData<List<DailyPushNotificationDto>> listLiveData;

	private DailyNotificationHelper dailyNotificationHelper;
	private DailyPushNotificationDto savedNotificationDto;
	private DailyPushNotificationDto newNotificationDto;
	private DailyPushNotificationDto editingNotificationDto;

	private boolean newNotificationSession;
	private boolean selectedFavoriteLocation;

	private Bundle bundle;

	private WeatherProviderType mainWeatherProviderType;
	private FavoriteAddressDto selectedFavoriteAddressDto;

	public DailyNotificationViewModel(@NonNull Application application) {
		super(application);
		listLiveData = repository.listLiveData();
		dailyNotificationHelper = new DailyNotificationHelper(application.getApplicationContext());
	}

	public DailyNotificationHelper getDailyNotificationHelper() {
		return dailyNotificationHelper;
	}

	public DailyNotificationViewModel setDailyNotificationHelper(DailyNotificationHelper dailyNotificationHelper) {
		this.dailyNotificationHelper = dailyNotificationHelper;
		return this;
	}

	public DailyPushNotificationDto getSavedNotificationDto() {
		return savedNotificationDto;
	}

	public DailyNotificationViewModel setSavedNotificationDto(DailyPushNotificationDto savedNotificationDto) {
		this.savedNotificationDto = savedNotificationDto;
		return this;
	}

	public DailyPushNotificationDto getNewNotificationDto() {
		return newNotificationDto;
	}

	public DailyNotificationViewModel setNewNotificationDto(DailyPushNotificationDto newNotificationDto) {
		this.newNotificationDto = newNotificationDto;
		return this;
	}

	public DailyPushNotificationDto getEditingNotificationDto() {
		return editingNotificationDto;
	}

	public DailyNotificationViewModel setEditingNotificationDto(DailyPushNotificationDto editingNotificationDto) {
		this.editingNotificationDto = editingNotificationDto;
		return this;
	}

	public boolean isNewNotificationSession() {
		return newNotificationSession;
	}

	public void setNotificationSession(boolean newNotificationSession) {
		this.newNotificationSession = newNotificationSession;
		if (newNotificationSession) {
			newNotificationDto = new DailyPushNotificationDto();
			newNotificationDto.setEnabled(true);
			newNotificationDto.setAlarmClock(LocalTime.of(8, 0).toString());
			newNotificationDto.addWeatherSourceType(getMainWeatherProviderType());
			newNotificationDto.setNotificationType(DailyPushNotificationType.First);
			newNotificationDto.setLocationType(LocationType.CurrentLocation);

			editingNotificationDto = newNotificationDto;
		} else {
			savedNotificationDto = (DailyPushNotificationDto) bundle.getSerializable("dto");
			editingNotificationDto = savedNotificationDto;

			if (editingNotificationDto.getLocationType() == LocationType.SelectedAddress) {
				selectedFavoriteAddressDto = new FavoriteAddressDto();
				selectedFavoriteAddressDto.setDisplayName(editingNotificationDto.getAddressName());
				selectedFavoriteAddressDto.setCountryCode(editingNotificationDto.getCountryCode());
				selectedFavoriteAddressDto.setLatitude(String.valueOf(editingNotificationDto.getLatitude()));
				selectedFavoriteAddressDto.setLongitude(String.valueOf(editingNotificationDto.getLongitude()));

				selectedFavoriteLocation = true;
			}
		}
	}

	public boolean isSelectedFavoriteLocation() {
		return selectedFavoriteLocation;
	}

	public DailyNotificationViewModel setSelectedFavoriteLocation(boolean selectedFavoriteLocation) {
		this.selectedFavoriteLocation = selectedFavoriteLocation;
		return this;
	}

	public Bundle getBundle() {
		return bundle;
	}

	public DailyNotificationViewModel setBundle(Bundle bundle) {
		this.bundle = bundle;
		return this;
	}

	public WeatherProviderType getMainWeatherProviderType() {
		return mainWeatherProviderType;
	}

	public DailyNotificationViewModel setMainWeatherProviderType(WeatherProviderType mainWeatherProviderType) {
		this.mainWeatherProviderType = mainWeatherProviderType;
		return this;
	}

	public FavoriteAddressDto getSelectedFavoriteAddressDto() {
		return selectedFavoriteAddressDto;
	}

	public DailyNotificationViewModel setSelectedFavoriteAddressDto(FavoriteAddressDto selectedFavoriteAddressDto) {
		this.selectedFavoriteAddressDto = selectedFavoriteAddressDto;
		return this;
	}

	public void getAll(DbQueryCallback<List<DailyPushNotificationDto>> callback) {
		repository.getAll(callback);
	}

	public void size(DbQueryCallback<Integer> callback) {
		repository.size(callback);
	}

	public void get(int id, DbQueryCallback<DailyPushNotificationDto> callback) {
		repository.get(id, callback);
	}

	public void delete(DailyPushNotificationDto dailyPushNotificationDto, @Nullable DbQueryCallback<Boolean> callback) {
		repository.delete(dailyPushNotificationDto, callback);
	}

	public void add(DailyPushNotificationDto dailyPushNotificationDto, DbQueryCallback<DailyPushNotificationDto> callback) {
		repository.add(dailyPushNotificationDto, callback);
	}

	public void update(DailyPushNotificationDto dailyPushNotificationDto, @Nullable DbQueryCallback<DailyPushNotificationDto> callback) {
		repository.update(dailyPushNotificationDto, callback);
	}

}

File: app/src/main/java/com/lifedawn/bestweather/ui/notification/daily/viewmodel/DailyNotificationViewModel.kt
package com.lifedawn.bestweather.ui.notification.daily.viewmodel

import android.app.Application
import android.os.Bundle
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import com.lifedawn.bestweather.data.local.room.callback.DbQueryCallback
import java.time.LocalTime

class DailyNotificationViewModel constructor(application: Application) : AndroidViewModel(application) {
    private val repository: DailyPushNotificationRepository = DailyPushNotificationRepository.getINSTANCE()
    val listLiveData: LiveData<List<DailyPushNotificationDto>>
    private var dailyNotificationHelper: DailyNotificationHelper
    private var savedNotificationDto: DailyPushNotificationDto? = null
    private var newNotificationDto: DailyPushNotificationDto? = null
    private var editingNotificationDto: DailyPushNotificationDto? = null
    var isNewNotificationSession: Boolean = false
        private set
    var isSelectedFavoriteLocation: Boolean = false
        private set
    var bundle: Bundle? = null
        private set
    private var mainWeatherProviderType: WeatherProviderType? = null
    private var selectedFavoriteAddressDto: FavoriteAddressDto? = null

    init {
        listLiveData = repository.listLiveData()
        dailyNotificationHelper = DailyNotificationHelper(application.getApplicationContext())
    }

    fun getDailyNotificationHelper(): DailyNotificationHelper {
        return dailyNotificationHelper
    }

    fun setDailyNotificationHelper(dailyNotificationHelper: DailyNotificationHelper): DailyNotificationViewModel {
        this.dailyNotificationHelper = dailyNotificationHelper
        return this
    }

    fun getSavedNotificationDto(): DailyPushNotificationDto? {
        return savedNotificationDto
    }

    fun setSavedNotificationDto(savedNotificationDto: DailyPushNotificationDto?): DailyNotificationViewModel {
        this.savedNotificationDto = savedNotificationDto
        return this
    }

    fun getNewNotificationDto(): DailyPushNotificationDto? {
        return newNotificationDto
    }

    fun setNewNotificationDto(newNotificationDto: DailyPushNotificationDto?): DailyNotificationViewModel {
        this.newNotificationDto = newNotificationDto
        return this
    }

    fun getEditingNotificationDto(): DailyPushNotificationDto? {
        return editingNotificationDto
    }

    fun setEditingNotificationDto(editingNotificationDto: DailyPushNotificationDto?): DailyNotificationViewModel {
        this.editingNotificationDto = editingNotificationDto
        return this
    }

    fun setNotificationSession(newNotificationSession: Boolean) {
        isNewNotificationSession = newNotificationSession
        if (newNotificationSession) {
            newNotificationDto = DailyPushNotificationDto()
            newNotificationDto.isEnabled = true
            newNotificationDto.alarmClock = LocalTime.of(8, 0).toString()
            newNotificationDto.weatherProviderType = getMainWeatherProviderType()
            newNotificationDto.notificationType = DailyPushNotificationType.First
            newNotificationDto.locationType = LocationType.CurrentLocation
            editingNotificationDto = newNotificationDto
        } else {
            savedNotificationDto = bundle!!.getSerializable("dto") as DailyPushNotificationDto?
            editingNotificationDto = savedNotificationDto
            if (editingNotificationDto.locationType === LocationType.SelectedAddress) {
                selectedFavoriteAddressDto = FavoriteAddressDto()
                selectedFavoriteAddressDto.displayName = editingNotificationDto.addressName
                selectedFavoriteAddressDto.countryCode = editingNotificationDto.countryCode
                selectedFavoriteAddressDto.latitude = editingNotificationDto.latitude.toString()
                selectedFavoriteAddressDto.longitude = editingNotificationDto.longitude.toString()
                isSelectedFavoriteLocation = true
            }
        }
    }

    fun setSelectedFavoriteLocation(selectedFavoriteLocation: Boolean): DailyNotificationViewModel {
        isSelectedFavoriteLocation = selectedFavoriteLocation
        return this
    }

    fun setBundle(bundle: Bundle?): DailyNotificationViewModel {
        this.bundle = bundle
        return this
    }

    fun getMainWeatherProviderType(): WeatherProviderType? {
        return mainWeatherProviderType
    }

    fun setMainWeatherProviderType(mainWeatherProviderType: WeatherProviderType?): DailyNotificationViewModel {
        this.mainWeatherProviderType = mainWeatherProviderType
        return this
    }

    fun getSelectedFavoriteAddressDto(): FavoriteAddressDto? {
        return selectedFavoriteAddressDto
    }

    fun setSelectedFavoriteAddressDto(selectedFavoriteAddressDto: FavoriteAddressDto?): DailyNotificationViewModel {
        this.selectedFavoriteAddressDto = selectedFavoriteAddressDto
        return this
    }

    fun getAll(callback: DbQueryCallback<List<DailyPushNotificationDto?>?>?) {
        repository.getAll(callback)
    }

    fun size(callback: DbQueryCallback<Int?>?) {
        repository.size(callback)
    }

    operator fun get(id: Int, callback: DbQueryCallback<DailyPushNotificationDto?>?) {
        repository.get(id, callback)
    }

    fun delete(dailyPushNotificationDto: DailyPushNotificationDto?, callback: DbQueryCallback<Boolean?>?) {
        repository.delete(dailyPushNotificationDto, callback)
    }

    fun add(dailyPushNotificationDto: DailyPushNotificationDto?, callback: DbQueryCallback<DailyPushNotificationDto?>?) {
        repository.add(dailyPushNotificationDto, callback)
    }

    fun update(dailyPushNotificationDto: DailyPushNotificationDto?, callback: DbQueryCallback<DailyPushNotificationDto?>?) {
        repository.update(dailyPushNotificationDto, callback)
    }
}